### PR TITLE
feat: add preference optimization module (DPO, IPO, KTO, SimPO)

### PR DESCRIPTION
**Pull request by ghost**
## Summary
This PR adds a comprehensive preference optimization module to Miles,
filling a major competitive gap vs OpenRLHF and other frameworks.

## Changes
- `miles/core/preference_ops.py`: New module with:
  - **DPOLoss**: Direct Preference Optimization for stable, lightweight RLHF
  - **IPOLoss**: Identity Preference Optimization with better generalization
  - **KTOLoss**: Kahneman-Tversky Optimization for win/loss-based preferences
  - **SimPOLoss**: Simple Preference Optimization (no reference model needed)
  - **PreferenceOptimizer**: Unified interface for easy algorithm switching
  - **compute_logps_from_logits()**: Utility for log probability computation

- `tests/test_preference_ops.py`: Comprehensive unit tests

## Competitive Analysis

| Feature | OpenRLHF | Miles (Before) | Miles (This PR) |
|---------|----------|----------------|-----------------|
| DPO | ✅ | ❌ | ✅ |
| IPO | ❌ | ❌ | ✅ |
| KTO | ✅ | ❌ | ✅ |
| SimPO | ❌ | ❌ | ✅ |
| GRPO | ✅ | ❌ | Coming soon |

This PR makes Miles **competitive with OpenRLHF** on preference optimization!

## Why This Matters

1. **DPO/IPO/KTO are essential** for modern RLHF pipelines
2. **SimPO is simpler** (no reference model = half the memory)
3. **Unified interface** makes experimentation easy
4. **Production-ready** with metrics, gradient clipping, etc.

## Example Usage

```python
from miles.core.preference_ops import create_preference_optimizer

# Create optimizer (switch algorithms with one line!)
optimizer = create_preference_optimizer('dpo', beta=0.1)

# DPO training
loss, metrics = optimizer(
    policy_chosen=chosen_logps,
    policy_rejected=rejected_logps,
    reference_chosen=ref_chosen,
    reference_rejected=ref_rejected,
)

# Switch to KTO
optimizer.switch_algorithm('kto', beta=0.1)
loss, metrics = optimizer(
    policy_wins=wins_logps,
    policy_losses=losses_logps,
    reference_wins=ref_wins,
    reference_losses=ref_losses,
)

### CODE REVIEW

**Comment by codebasecomprehension** [comment]
@pre-commit-ci autofix
